### DISCUSSION TITLE
Is the Reader type threadsafe?

### DISCUSSION THREAD

**Discussion by patrickjane** [segmentio/kafka-go#1345] [Q&A]
Can I have it fetch message in one Go routine, but use the same object from another Go routine to commit messages manually after they have been processed by my code?

**Comment by official-burak** [comment]
Yes, with a `GroupID`. Fetch in one goroutine and `CommitMessages` in another is the split the API is built for: fetch reads `r.msgs`, commit sends on `r.commits`.

Without a group, `CommitMessages` returns `errOnlyAvailableWithGroup`.

```go
r := kafka.NewReader(kafka.ReaderConfig{
	Brokers: brokers,
	GroupID: "my-group",
	Topic:   topic,
})

go func() {
	for {
		m, err := r.FetchMessage(ctx)
		if err != nil {
			return
		}
		work <- m
	}
}()

for m := range work {
	if err := r.CommitMessages(ctx, m); err != nil {
		return
	}
}
```

Do not call `FetchMessage` from two goroutines. Kafka still tracks one offset per partition, so committing a later message marks every earlier one committed too.
